### PR TITLE
Scrape results from Serper results using Olostep

### PR DESCRIPTION
**Pull request by sabaimran**
- Use [Olostep](https://www.olostep.com/) to go through top results from Serper's google search
- Add an intermediate step that summarizes extracted web results for further response downstream
- If `answerBox` is present in the query results, no need to do the additional scraping
- Add truncation logic to the model wrapper

### CODE REVIEW

File: src/khoj/processor/conversation/utils.py [hunk 1]
```diff
@@ -194,6 +196,7 @@ def truncate_messages(
         assert type(system_message.content) == str
         current_message = "\n".join(messages[0].content.split("\n")[:-1]) if type(messages[0].content) == str else ""
         original_question = "\n".join(messages[0].content.split("\n")[-1:]) if type(messages[0].content) == str else ""
+        original_question = f"\n{original_question}"
```
Comment (debanjum) [inline src/khoj/processor/conversation/utils.py:199]: Why add a new line before the `original_question`?

File: src/khoj/routers/helpers.py [hunk 2]
```diff
@@ -175,35 +195,41 @@ async def generate_better_image_prompt(q: str, conversation_history: str) -> str
...
             state.gpt4all_processor_config = GPT4AllProcessorModel(conversation_config.chat_model)
 
         loaded_model = state.gpt4all_processor_config.loaded_model
         return send_message_to_model_offline(
-            message=message,
+            message=truncated_messages[-1].content,
             loaded_model=loaded_model,
             model=conversation_config.chat_model,
             streaming=False,
+            system_message=truncated_messages[0].content,
```
Comment (debanjum) [inline src/khoj/routers/helpers.py:219]: Is system message always the first of the `truncated_messages`? Why not just set system_message to system_message?

```suggestion
            system_message=system_message,
```

**Comment by debanjum** [comment]
Can we add chat actor, director tests to verify the functioning of this /online with website crawl flows? It's getting a little complicated. It'd be good to see what kind of stuff Khoj can do with this vs without it (It'll improve insight into our chatquality benchmarking via pytests)

File: src/khoj/processor/tools/online_search.py [hunk 3]
```diff
@@ -51,4 +70,41 @@ def _search_with_google(subquery: str):
...
+    for subquery in response_dict:
+        # If a high quality answer is not found, search the web pages of the first 3 organic results
+        if is_none_or_empty(response_dict[subquery].get("answerBox")):
+            extracted_content[subquery] = []
+            for result in response_dict[subquery].get("organic")[:1]:
+                logger.info(f"Searching web page of '{result['link']}'")
+                try:
+                    extracted_content[subquery].append(search_with_olostep(result["link"]).strip())
+                except Exception as e:
+                    logger.error(e, exc_info=True)
```
Comment (debanjum) [inline src/khoj/processor/tools/online_search.py:87]: Maybe good to prefix a description of the error for context
```suggestion
                    logger.error(f'Failed to use Olostep to search through {result["link"]}.\nException: {e}", exc_info=True)
```

**Review by debanjum — APPROVED** [review]
This change will be super useful for deeper insights from online results! It'd be great if we can add some chat quality tests to evaluate Khoj's improved capabilities with these changes

File: src/khoj/processor/conversation/utils.py (hunk 1 above)
Comment (sabaimran) [reply src/khoj/processor/conversation/utils.py:199]: It's getting concatenated at the end without a new line if it's not explicitly added, which is why I'm adding it in here (before truncation).

File: src/khoj/routers/helpers.py (hunk 2 above)
Comment (sabaimran) [reply src/khoj/routers/helpers.py:219]: We would probably want to leave an option for the truncation logic to act as an underlying abstraction layer to handle longer messages. I wouldn't want to override its decision if/when we apply some truncation logic to the system prompt.

**Pull request merged**